internal/events/tokens_transferred.go

```diff
@@ -103,10 +103,17 @@ func (em *eventManager) persistTokenTransfer(ctx context.Context, transfer *toke
 	em.emitBlockchainEventMetric(transfer.Event)
 	transfer.BlockchainEvent = chainEvent.ID
 
-	if err := em.database.UpsertTokenTransfer(ctx, &transfer.TokenTransfer); err != nil {
+	existing, err := em.database.InsertUpdateOrGetTokenTransfer(ctx, &transfer.TokenTransfer)
```

Comment:
I agree that I think you should be able to remove the earlier `GetTokenTransferByProtocolID` and rely only on this call to find duplicates. The earlier inserts for the transaction and blockchain event should be idempotent and replayable in and of themselves.

It could cause an issue I'm not thinking of now... but I don't see any way that could cause an issue with websocket acks like [what you were describing](https://github.com/hyperledger/firefly/pull/1291#issuecomment-1525735229). Perhaps it could be a coincidence, and that is a separate issue?